test: tx orphan handling

glozow opened the pull request:
I was doing some mutation testing (through reckless refactoring) locally and found some specific behaviors in orphan handling that weren't picked up by tests. Adding some of these test cases now can maybe help with reviewing refactors like #28031.

- Parent requests aren't sent immediately. A delay is added and the requests are filtered by AlreadyHaveTx before they are sent, which means you can't use fake orphans to probe precise arrival timing of a tx.
- Parent requests include all that are not AlreadyHaveTx. This means old confirmed parents may be requested.
- The node does not give up on orphans if the peer responds to a parent request with notfound. This means that if a parent is an old confirmed transaction (in which notfound is expected), the orphan should still be resolved.
- Rejected parents can cause an orphan to be dropped, but it depends on the reason and only based on txid.
- Rejected parents can cause an orphan to be rejected too, by both wtxid and txid.
- Requests for orphan parents should be de-duplicated with "regular" txrequest. If a missing parent has the same hash as an in-flight request, it shouldn't be requested.
- Multiple orphans with overlapping parents should not cause duplicated parent requests.

glozow committed:
[functional test] transaction orphan handling

dergoegge commented:
Concept ACK

Good to have these tests prior to refactoring

jamesob commented:
Concept ACK, looks like some great additional coverage.

brunoerg commented:
Concept ACK

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        mempool_tx = self.wallet.send_self_transfer(from_node=node)
+        utxo_unconf_mempool = mempool_tx["new_utxo"]
+
+        # This UTXO is unconfirmed and missing.
```
took me a second, just assert it's not in mempool for those quickly reading

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        self.wait_until(test_function, timeout=10)
+
+    def assert_message_ignored(self, message):
+        """Check that the node does not respond to this message with any of
```
```suggestion
        """Check that the node does not immediately respond to this message with any of
```

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        # Spy peer should not be able to query the node for the parent yet, since it hasn't been
+        # announced / insufficient time has elapsed.
+        parent_inv = CInv(t=MSG_WTX, h=int(tx_real["tx"].getwtxid(), 16))
+        peer_spy.assert_message_ignored(msg_getdata([parent_inv]))
```
can we assert that peer_spy hasn't received the INV, just to be sure?

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        self.fastforward(TXID_RELAY_DELAY)
+        peer_spy.sync_with_ping()
+
+        self.log.info("Test parent requests don't reveal whether the parent was present when the orphan arrived")
```
I don't get this case. Why wouldn't you request a "fake" parent if you don't know it's fake already?

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        # Parent of child_A only
+        missing_parent_A = self.wallet_nonsegwit.create_self_transfer()
+        # Parents of child_A and child_B
+        missing_parent_AB = self.wallet_nonsegwit.create_self_transfer()
```
does this wallet never use 0-conf change? would be nice to assert to make it clear the lack of utxo connection

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        self.relay_transaction(peer_orphans, child_A["tx"])
+        self.fastforward(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        # Both missing parents should be requested.
+        peer_orphans.wait_for_parent_requests([int(missing_parent_A["txid"], 16), int(missing_parent_AB["txid"], 16)])
```
"The node should not request a parent if it has an in-flight txrequest" ? Seems like parents are being reqiested. Maybe a typo or I can't tell what scenario it's covering.

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,439 @@
+        peer2 = node.add_p2p_connection(PeerTxRelayer())
+        peer3 = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test orphan handling when nonsegwit parent paid 0 fee")
```
bit of a repeat with test_orphan_rejected_parents_exceptions ?

glozow replied:
(I didn't change it but I also don't feel strongly, could combine if you want)

instagibbs replied:
nope it's fine

instagibbs reviewed (commented):
moar coverage good

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,424 @@
+        # Node receives transaction. It attempts to obfuscate the exact timing at which this
+        # transaction entered its mempool. Send unsolicited because otherwise we need to wait for
+        # request delays.
+        peer_normal.send_and_ping(msg_tx(tx_parent_arrives["tx"]))
```
```suggestion
        peer_normal.send_and_ping(msg_tx(tx_parent_arrives["tx"]))
        assert tx_parent_arrives["txid"] in node.getrawmempool()
```
my brain for some reason was thinking it was in the orphan pool as well

Jones098 replied:
Release

glozow replied:
> Should we add an rpc to allow querying the orphanage?

Could do that, as there's a bit of black box-ness here. Though I also think that the ideal situation would be to write unit tests if we're interested in the exact contents of the orphanage.

Empact commented:
Concept ACK

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/28199/commits/abe8536192c9f2cd6ba9d0e083f23dec4d20841f

https://github.com/bitcoin/bitcoin/pull/28199#discussion_r1283271742 would be nice to be slightly cleaned up for future readers

dergoegge reviewed (approved):
Code review ACK abe8536192c9f2cd6ba9d0e083f23dec4d20841f

ajtowns commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,424 @@
+        peer1 = node.add_p2p_connection(PeerTxRelayer())
+        peer2 = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test orphan handling when nonsegwit parent paid 0 fee")
```
Might be better to be clearer about the different behaviours you're testing for, rather than how you're triggering those behaviours? eg:

 * Orphan handling when parent is known to be invalid
 * Orphan handling when segwit parent may be retried with alternate witness data

glozow replied:
Agree, updated the logging

ajtowns commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,424 @@
+        child = self.wallet.create_self_transfer(utxo_to_spend=parent['new_utxo'])
+        return child["tx"].getwtxid(), child["tx"], parent["tx"]
+
+    def fastforward(self, seconds):
```
Maybe name this something that's a bit more clearly related to mocktime (`bumpmocktime`?) and add it to `BitcoinTestFramework` ?

glozow replied:
Good point, added a `bumpmocktime` to test framework, I'll probably use it in other tests in the future.

ajtowns reviewed (commented):
ACK

I think we should probably prefer to test behaviours that are intentional/desirable, rather than behaviours that just happen to be how it got implemented. It's not immediately clear to me which class the things being tested here fall into.

glozow reviewed (commented):
> I think we should probably prefer to test behaviours that are intentional/desirable, rather than behaviours that just happen to be how it got implemented.

Yeah the purpose of writing these tests was largely documentation and not changing behavior unintentionally.

I think most of the things here are intentional, though can think of at least one thing that seems to be "that's just how it got implemented." I was trying to figure out whether we would want to use a `notfound` for a parent tx to mean we should give up on processing the orphan that depends on it. Currently we don't do anything since we don't track the fact that the request is related to an orphan. But since we just request based on `!AlreadyHaveTx` we might reject due to a long-confirmed parent. I suppose it would be different if validation told net_processing which inputs were missing and request those. Though that doesn't seem worth implementing if we can just do ancestor package relay.

glozow committed:
[test framework] make it easier to fast-forward setmocktime

Have each TestNode keep track of the last timestamp it called
setmocktime with, and add a bumpmocktime() function to bump by a
number of seconds. Makes it easy to fast forward n seconds without
keeping track of what the last timestamp was.

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,424 @@
+        node = self.nodes[0]
+        tx_parent_arrives = self.wallet.create_self_transfer()
+        tx_parent_doesnt_arrive = self.wallet.create_self_transfer()
+        # Fake orphan spends nonexistent outputs of the two parents
```
Sorree! Added explanation in the comment.

dergoegge commented on test/functional/test_framework/test_framework.py:
```diff
@@ -103,6 +103,11 @@ def __init__(self) -> None:
+        # setmocktime, it should call bumpmocktime(0) at the beginning of run_test in order to
+        # prevent spurious failures. If the test does not want to use setmocktime, it must not call
+        # setmocktime because the time will not move forward afterwards.
+        self.mocktime = int(time.time())
```
Wouldn't it make more sense to have `bumpmocktime` on TestNode and set `self.mocktime` when ever `setmocktime` is called?

* allows for node independent mocktime bumping
* allows to use an initial mocktime other than time.time()

glozow replied:
Good point, moved from TestFramework to TestNode

glozow replied:
Added a wrapper for the RPC to update `TestNode::mocktime` every time `setmocktime` is called.

dergoegge commented on test/functional/test_framework/test_node.py:
```diff
@@ -144,6 +144,11 @@ def __init__(self, i, datadir, *, chain, rpchost, timewait, timeout_factor, bitc
+        # The time doesn't move forwards after we mock it, so don't call setmocktime if the test
+        # doesn't intend to use it.  If a test wants to use setmocktime, it should call
+        # bumpmocktime(0) at the beginning of run_test in order to prevent spurious failures.
+        self.mocktime = int(time.time())
```
Sorry for being nitty about this but it's a little weird to have this set to some time that is different from the actual mock time of the node.

I would suggest initializing this to `None`, resetting it to `None` if `setmocktime(0)` was called and additionally assert that it is not None in `bumpmocktime`.

This ensures it is not set to some differing value (different from the actual mocktime) and makes sure `bumpmocktime` is used correctly. `TestNode.mocktime` can then also be reliably used as a getter for the node's mocktime.

glozow replied:
Done. I like that the assertion forces tests to make an initial setmocktime call before calling bumpmocktime.

dergoegge reviewed (approved):
reACK 9eac5a0529f869075f0331e40d322c34fc8fc2af

achow101 commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer.sync_with_ping()
+        assert_equal(len(peer.last_message["getdata"].inv), 2)
+        peer.wait_for_parent_requests([int(txid_conf_old, 16), int(missing_tx["txid"], 16)])
```
Requesting the already confirmed transaction is surprising to me. Is this expected to change in future work (if so a comment mentioning that would be useful) or is it intended behavior?

glozow replied:
> Requesting the already confirmed transaction is surprising to me.

Yep, was strange to me too, but `AlreadyHaveTx` is peerman's best metric for determining what the missing parents are (validation doesn't say which one(s) are missing). Also afaiu it's considered uncommon to (1) have an orphan (2) have a transaction that spends a very old coin, so I'm guessing this is quite rare.

> Is this expected to change in future work

I'm not aware of any plans to change the legacy way of requesting orphan parents. But with package relay we'd ask for the list of unconfirmed ancestors, so we wouldn't have this kind of false positive.

achow101 commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+        assert grandchild["txid"] != grandchild["tx"].getwtxid()
+
+        # Relay the parent. It should be rejected because it pays 0 fees.
+        self.relay_transaction(peer1, parent_low_fee_nonsegwit["tx"])
```
Could check that `parent_low_fee_nonsegwit` is not in the mempool as done in the other test cases.

achow101 reviewed (commented):
ACK 9eac5a0529f869075f0331e40d322c34fc8fc2af

fjahr commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+        peer.wait_for_parent_requests([int(missing_grandparent["txid"], 16)])
+
+        # The node should put the orphan into the orphanage and request missing_parent, skipping
+        # missing_parent_orphan because it already has it in the orphanage.
```
Could add another assertion here that `missing_parent_orphan` is indeed only requested once.

fjahr commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+
+        confirmed_utxos = [self.wallet_nonsegwit.get_utxo() for _ in range(4)]
+        assert all([utxo["confirmations"] > 0 for utxo in confirmed_utxos])
+        self.log.info("Test handling of multiple orphans with missing parents that are already being requested")
```
nit: I would argue that the (or the first) log should always be the first line of the test. Otherwise it may be confusing if an error happens in the code above it may look like the previous test failed.

fjahr commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        # There are 3 missing parents. missing_parent_A and missing_parent_AB should be requested.
+        # But inflight_parent_AB should not, because there is already an in-flight request for it.
+        peer_orphans.wait_for_parent_requests([int(missing_parent_A["txid"], 16), int(missing_parent_AB["txid"], 16)])
```
nit: I was a bit irritated that that `inflight_parent_AB` not being requested isn't checked here but I see it follows below. For expressiveness you could consider duplicating the check here or adding a small comment.

fjahr commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+
+        self.log.info("Test that an orphan with rejected parents, along with any descendants, cannot be retried with an alternate witness")
+        parent_low_fee_nonsegwit = self.wallet_nonsegwit.create_self_transfer(fee_rate=0)
+        assert_equal(parent_low_fee_nonsegwit["txid"], parent_low_fee_nonsegwit["tx"].getwtxid())
```
nit: This sanity check is repeated multiple times. I think it would be enough to do it once when setting up the `MiniWallet` since that seems to be the only way there could be an issue, if somehow MiniWallet is broken.

fjahr reviewed (commented):
Code review ACK 9eac5a0529f869075f0331e40d322c34fc8fc2af

...modulo removing the dead code. Otherwise mostly nit-ish comments, I can also open a follow-up myself if someone feels like merging this as is.

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -0,0 +1,416 @@
+        self.num_nodes = 1
+        self.extra_args = [[]]
+
+    def create_parent_and_child(self):
```
My bad!

> I can also open a follow-up myself if someone feels like merging this as is.

If it helps, they were written to be used in #28031. I pulled these tests out to merge before the refactors and I guess forgot to remove the helpers that aren't used (yet).

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/28199/commits/9eac5a0529f869075f0331e40d322c34fc8fc2af
